### DISCUSSION TITLE
Which BlockIDFlag is used when a validator does not vote?

### DISCUSSION THREAD

**Discussion by rootulp** [tendermint/tendermint#9557] [Q&A] [answered]
I have a question regarding the [BlockIDFlag](https://github.com/tendermint/tendermint/blob/master/spec/core/data_structures.md#blockidflag). A section in the specs say:

> Represents the validators participation in consensus: Either voted for the block that received the majority, voted for another block, voted nil or did not vote

Which BlockIDFlag is used when a validator does not vote? `BLOCK_ID_FLAG_UNKNOWN` or `BLOCK_ID_FLAG_ABSENT`? I assume the later based on the name but then it's not clear to me when the former is used.

**Comment by sergio-mena** [comment, marked as answer]
Hi @rootulp, 
Indeed, we use `BLOCK_ID_FLAG_ABSENT` to signal that the validator's vote was not received.
`BLOCK_ID_FLAG_UNKNOWN` was introduced into the protobuf version of this type to make sure it is always filled up. IOW, a protobuf value of `BLOCK_ID_FLAG_UNKNOWN` is incorrect: it automatically indicates there is an error condition.
Further, notice that the [corresponding domain type](https://github.com/tendermint/tendermint/blob/main/types/block.go#L576-L585), does not have "unknown", only "absent", "commit", and "nil"

**Reply by rootulp** [reply]
Thanks for the super helpful answer!
